### ISSUE TITLE
Support proprietary/individual licenses (SPDX license references/magic numbers next to SPDX License IDs)

### ISSUE THREAD

**Issue by ktomk** [composer/composer#2152]
*Mike Graf* [asked Jan 2013](https://groups.google.com/forum/#!topic/composer-users/_O0hYtltI2Y) in the composer-users mailing list:

> Any hints at the proper license identifier to use for "proprietary" or "closed source" ? I want `composer validate` to stop barking at me about the lack of a license.

He sees this warning because the composer validation is checking for valid *SPDX License IDs* and giving a warning if such an ID is not found. This was introduced via [PR #661 *Proposal to use SPDX license identifiers*](https://github.com/composer/composer/pull/661).

By fact, the [curated SPDX License List](https://spdx.org/licenses/) does not contain an ID for a license that is not further specified (e.g. like a placeholder). Also one of the [main purposes of this license list (SPDX-LL)](http://spdx.org/spdx-license-list/license-list-overview) is to identify common FLOSS licenses easily (hence representing them by a short identifier (ID)).

As we did suggest to use identifiers from that list earlier, the use-case using a (not further specified) proprietary license of *Mike Graf* was not covered. Giving warnings in validation did reveal this needed but unsupported use-case.

*Jordi Boggiano* suggested to use a *special ID* (*Special Case Value*; *Magic Number*) `proprietary` and implemented that in https://github.com/composer/composer/commit/e43d0b5a5b8dec50464c907bf837e673f50e1de2

This change looks like some kind of "quick fix" to me, as it covers the concrete use-case but I was not able to find any kind of further reference which adds more context to this change, e.g. if this is a common way to use such a string.

Me as the original implementer was not aware of that report of the shortcoming until this last weekend and I started to review it. Following is my discussion / analysis and suggestion how to change composer for better individual license support.

----

The SPDX License list does not have an ID for an unspecified license so there is no concrete ID to suggest.

However, looking further the [SPDX 1.1. Specification](http://spdx.org/sites/spdx/files/spdx-1.1.pdf) (Aug 2012; in the following section numbers related to this revision of the document) this license list is an appendix of (*Appendix I*) suggests for comparable properties/fields like *Concluded License* (4.10) the following *Data-Format* (4.10.4):

>     Data Format:    <short form identifier in Appendix I> | "LicenseRef"-N | “NOASSERTION”
                      | “NONE” | <license set>

The `<short form identifier in Appendix I>` (a) stands for the License IDs used so far. The `<license set>` inherits as known and already used by composer (disjunctive / conjunctive list, see http://getcomposer.org/doc/04-schema.md#license). The overall meaning is explained in 4.10:

> - (a) the SPDX License List short form identifier, if the concluded license is on the SPDX License 
List;
> - (b) a reference to the license text denoted by the LicenseRef-#, if the concluded license is not on the SPDX License List;
> - (c) NOASSERTION  should be used if:  
>  - (i) the SPDX file creator has attempted to but cannot reach a reasonable objective 
determination of the Concluded License;  
>  - (ii) the SPDX file creator is uncomfortable concluding a license, despite some license 
information being available; 
>  - (iii) the SPDX file creator has made no attempt to determine a Concluded License; 
>  - (iv) the SPDX file creator has intentionally provided no information (no meaning should 
be implied by doing so); or
>  - (v) there is no licensing information from which to conclude a license for the package.

Re-using the principles in here, it seems appropriate that most often `NOASSERTION` should be used in case the license information is not provided. By the feedback given in the mailing-list, the license information is not provided because there is no ID hence the license is not asserted / undefined. The meaning is equally well to the suggested `proprietary` as it is as well not giving an assertion about the concrete license nor is the license information provided (in the thread it has been missing to offer more information like a concrete license text so it remains undefined whether or not there is some license text actually, nor was there feedback after the change-set which introduced `proprietary` so far). However using `NOASSERTION` has the benefit that the string is most likely reserved within SPDX as it has a special meaning *already*. This is a first suggestion of mine.

The `"LicenseRef"-N` format does allow to specify an individual license with the package. So with the circumstance that the proprietary license is actually available, it could be added as a file in the package root like named the same `LicenseRef-1` and then the JSON data can point to it. This has the benefit that a license is actually specified (and like in SPDX there is no need to have one from the curated license list but to include the own one). This is a second suggestion of mine how to deal with this, I will extend it later on as I do not really like the "external" file solution. However this extends the change so far to actually specify a concrete license text not part of the list.

The `"NONE"` format should be supported as well because it's used in SPDX. Likewise for all other formats validating license specifiers. So the validation class can be modified to support those formats. Even `"NONE"` is suggested in the data-format under 4.10 it remains undefined in that section. Further on in 4.11 an explanation is given (same as well with 4.12 ...):

> [...] (c) NONE, if no license information is detected in any of the files; [...]

`"NONE"` is therefore reserved for the case that no licensing information is given. It makes sense to support this as well in my eyes.

Conclusion: Even the license identifier list does not allow to specify your own license, SPDX itself does allow and has some format that can follow these.

----

The initial feature request in the mailing-list didn't include an analysis what SPDX offers under such circumstances of an individual or proprietary license and only suggested a place-holder value so far. Reviewing SPDX in the context of the request reveals that SPDX has already a place-holder concept:

 - `"LicenseRef"-N` to specify *n* different licenses for a software package, whereas the reference is *local* (it is in the same file).
 - `"NOASSERTION"` to specify that no further assertion of a concrete license is done, e.g. in case of a proprietary license which often is not necessary as those pacakages are mostly for private package management and not providing a license has been done *by intention* (see description of (c) `NOASSERTION` in 4.10).
 - `"NONE"` to specify that no license information at all has been given (unintentionally), whether a concrete license or some non-concrete license. E.g. somebody who packages a software was not able to determine the license of somebodies else software by the files/data provided (see description of (c) `NONE` in 4.11).

To benefit from the feature request and from what SPDX offers already some thoughtful conclusions should be drawn.

The `"LicenseRef"-N` identifier makes only sense if there is a way to provide the license information with the package - or even better - inside the `composer.json`. 

The most interoperable part would be an SPDX representation of the package encoded in JSON format. This should be technically possible however it is of most work to be done and we likely will not see any adoption so far (as this is the same as an SPDX file, it technically isn't necessary either, so why should somebody use it?).

A more easy and practical solution that comes to mind is to externalize those license references into same-named files in the packages root *or* for a same-named object property inside the `composer.json` file containing the license text. This would be easier to validate as well.

An example in SPDX of such a local reference and text is the `<extractedText>` element in 5.2.6. I would suggest a more compressed form here as this is not about extraction but one step before making SPDX extraction more easy:

    {
        "license": "LicenseRef-1",
        "license-ref-1": ""THE BEER-WARE LICENSE" (Revision 42): \n<phk@FreeBSD.ORG> wrote this file. As long as you retain this notice you \ncan do whatever you want \nwith this stuff. If we meet some day, and you think this stuff is worth it, you can buy me a beer in return Poul-Henning Kamp "
    }

This is a more or less straight forward adoption allowing the license-id to be the property, modified for lower-case letters as common with composer json

    LicenseRef-1    =>     license-ref-1

The license ID used in the `"license"` property is converted the following: The upper-case first letter "L" is replaced with lower-case "l" and the upper-case letter "R" is replaced with "-r" as it is in the middle of the word. This conversion is one-way only unless other constraints like `N` is always an unsigned number is met, however the ambiguity is generally low, so even unstable it is pretty stable under normal circumstances. Stability is not a show-stopper anyway as it does not need to be stable, the reference is one-way here starting from the `license` property value pointing to the license text property name.

The property with that name then contains the *license text* (preferable). As this is a text-field a composer.json author could also provide a weak reference here like *"Contact <legal@xample.com> for usage details in projects with third-party code."* or similar. So even not a concrete license, such a text does offer some benefit for a user of the package even if internally. However this might be an edge-case for reasons of simplicity, authors should probably if they don't have the license text use `NOASSERTION` instead of using a reference so to make a distinct difference between not providing the license information and providing the license. As this is about providing the license, this should be the license text. As license information should be given in a README anyway, it's probably a better place to leave notices in there instead of disturbing the JSON schema by multiple meanings.

This is not fully following SPDX standard (as SPDX has more properties for a license reference than only the license text) however `composer.json` is not there to follow it anyway 1:1. Instead this allows users to specify their own license with a similar convention as SPDX does in backward / forward compatible manner. And composer.json still leans on SPDX, so this is helpful for anybody who needs to review a packages license.

This `"LicenseRef"-N` suggestion makes changes to the `composer.json` schema necessary. And supporting `"LicenseRef"-N` this way is my third suggestion and it supersedes the earlier one to represent license texts as files (see above).

----

As right now `proprietary` is suggested as the *magic number* to signal a proprietary license (a license unlisted in the curated SPDX license list), I suggest to change this in the future because such a word is not part of SPDX and SPDX offers wording here. After reviewing this feature request in context of SPDX, `NOASSERTION` looks like a more appropriate way to communicate that the license is not specified. As we lean on SPDX, missing support for it should be added.

As the request was reported only once so far, I'd suggest to change this to `NOASSERTION` for users of `proprietary` giving an upgrade warning for some time on future validation. Also a new way how to specify the license text in concrete should be given and it should be explained when `NOASSERTION` or `NONE` could be used in the docs.

As I think it's also worth, I'll report back some feedback to SPDX about this issue here so to make SPDX more interoperable for adoptions like it is done with composer. I can imagine other package managers / package registries can benefit from the steps made here as well.

I'm in good faith that the suggested changes will improve licensing support with `composer.json`, completing not only for licenses that are not further specified but as well for license which texts exists and which are not part of the license list (and will most likely will never become part because they are proprietary).

So to overall better support custom licensing in `composer.json`.

I'll compile a pull request to reflect these suggestions in code.

**Comment by Seldaek** [comment]
I think we can close this at this point :)

**Comment by alcohol** [comment]
Feel free to check out [composer/spdx-licenses](/composer/spdx-licenses) though.

**Comment by ktomk** [comment]
all fine :)
